File: app/src/main/java/com/lskycity/androidtools/ui/FeedbackActivity.java
package com.lskycity.androidtools.ui;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.support.customtabs.CustomTabsIntent;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.text.style.UnderlineSpan;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.lskycity.androidtools.AppConstants;
import com.lskycity.androidtools.R;
import com.lskycity.androidtools.app.BaseActivity;
import com.lskycity.androidtools.app.ToolApplication;
import com.lskycity.androidtools.apputils.Feedback;

import org.json.JSONObject;

/**
 * collect feedback from user
 *
 * @author zhaofliu
 * @since 1/6/17
 */

public class FeedbackActivity extends BaseActivity implements TextWatcher, View.OnClickListener {

    private static final int MAX_TEXT_COUNT = 125;

    private EditText feedback;
    private TextView textCount;

    private long clickTime = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_feedback);
        feedback = (EditText) findViewById(R.id.feedback_content);
        feedback.addTextChangedListener(this);
        textCount = (TextView) findViewById(R.id.text_count);


        TextView checkNewVersion = (TextView) findViewById(R.id.check_new_version);

        //link the check user guide activity.
        String checkNewVersionString = getString(R.string.check_new_version);
        Spannable checkNewVersionSpannable = new SpannableString(checkNewVersionString);
        checkNewVersionSpannable.setSpan(new UnderlineSpan(), 0, checkNewVersionString.length(), Spannable.SPAN_INCLUSIVE_EXCLUSIVE);
        checkNewVersion.setText(checkNewVersionSpannable);

        checkNewVersion.setOnClickListener(this);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.activity_feedback, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == R.id.menu_send) {
            sendFeedback(feedback.getText().toString());
            return true;
        }
        return false;
    }


    private void sendFeedback(String content) {
        long currentTime = System.currentTimeMillis();
        if(currentTime-clickTime<2000) {
            return;
        }

        clickTime = currentTime;

        if(TextUtils.isEmpty(content.trim())) {
            Toast.makeText(FeedbackActivity.this, R.string.send_feedback_empty, Toast.LENGTH_LONG).show();
            return;
        }

        JsonObjectRequest request = new JsonObjectRequest(Request.Method.POST, AppConstants.FEEDBACK_URL, Feedback.obtain(this, content).toJSONObject(), new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject jsonObject) {
                Toast.makeText(FeedbackActivity.this, R.string.send_feedback_success, Toast.LENGTH_LONG).show();
                supportFinishAfterTransition();
            }
        },

        new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError volleyError) {
                Toast.makeText(FeedbackActivity.this, getString(R.string.send_feedback_fail)+", "+volleyError.getMessage(), Toast.LENGTH_LONG).show();

            }
        });

        ToolApplication.get().getRequestQueue().add(request);
        Toast.makeText(FeedbackActivity.this, R.string.sending_feedback, Toast.LENGTH_SHORT).show();

    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {

    }

    @SuppressLint("SetTextI18n")
    @Override
    public void afterTextChanged(Editable s) {
        textCount.setText(s.length()+"/125");
        if(s.length() == MAX_TEXT_COUNT) {
            textCount.setTextColor(Color.RED);
        } else {
            textCount.setTextColor(getResources().getColor(R.color.text_color));
        }
    }

    @Override
    public void onClick(View v) {
        if(v.getId() == R.id.check_new_version) {
            AppConstants.startUrlWithCustomTab(this, AppConstants.MAIN_PAGE_URL);
        }
    }
}

File: app/src/main/java/com/lskycity/androidtools/ui/FeedbackActivity.kt
package com.lskycity.androidtools.ui

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.text.*
import android.text.style.UnderlineSpan
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.lskycity.androidtools.AppConstants
import com.lskycity.androidtools.R
import com.lskycity.androidtools.app.BaseActivity
import com.lskycity.androidtools.app.ToolApplication
import com.lskycity.androidtools.apputils.Feedback

/**
 * collect feedback from user
 *
 * @author zhaofliu
 * @since 1/6/17
 */

class FeedbackActivity : BaseActivity(), TextWatcher, View.OnClickListener {

    private lateinit var feedback: EditText
    private lateinit var textCount: TextView

    private var clickTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_feedback)
        feedback = findViewById<View>(R.id.feedback_content) as EditText
        feedback.addTextChangedListener(this)
        textCount = findViewById<View>(R.id.text_count) as TextView


        val checkNewVersion = findViewById<View>(R.id.check_new_version) as TextView

        //link the check user guide activity.
        val checkNewVersionString = getString(R.string.check_new_version)
        val checkNewVersionSpannable = SpannableString(checkNewVersionString)
        checkNewVersionSpannable.setSpan(UnderlineSpan(), 0, checkNewVersionString.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
        checkNewVersion.text = checkNewVersionSpannable

        checkNewVersion.setOnClickListener(this)

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_feedback, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_send) {
            sendFeedback(feedback.text.toString())
            return true
        }
        return false
    }


    private fun sendFeedback(content: String) {
        val currentTime = System.currentTimeMillis()
        if (currentTime - clickTime < 2000) {
            return
        }

        clickTime = currentTime

        if (TextUtils.isEmpty(content.trim { it <= ' ' })) {
            Toast.makeText(this@FeedbackActivity, R.string.send_feedback_empty, Toast.LENGTH_LONG).show()
            return
        }

        val request = JsonObjectRequest(Request.Method.POST, AppConstants.FEEDBACK_URL, Feedback.obtain(this, content).toJSONObject(), Response.Listener {
            Toast.makeText(this@FeedbackActivity, R.string.send_feedback_success, Toast.LENGTH_LONG).show()
            supportFinishAfterTransition()
        },

                Response.ErrorListener { volleyError -> Toast.makeText(this@FeedbackActivity, getString(R.string.send_feedback_fail) + ", " + volleyError.message, Toast.LENGTH_LONG).show() })

        ToolApplication.get().requestQueue.add(request)
        Toast.makeText(this@FeedbackActivity, R.string.sending_feedback, Toast.LENGTH_SHORT).show()

    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

    }

    @SuppressLint("SetTextI18n")
    override fun afterTextChanged(s: Editable) {
        textCount.text = s.length.toString() + "/125"
        if (s.length == MAX_TEXT_COUNT) {
            textCount.setTextColor(Color.RED)
        } else {
            textCount.setTextColor(ResourcesCompat.getColor(resources, R.color.text_color, theme))
        }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.check_new_version) {
            AppConstants.startUrlWithCustomTab(this, AppConstants.MAIN_PAGE_URL)
        }
    }

    companion object {
        private val MAX_TEXT_COUNT = 125
    }
}
